### ISSUE TITLE
Ability to block one pipeline from starting while another is running; or, try/finally support

### ISSUE THREAD

**Issue by tillig** [spinnaker/spinnaker#4030]
### Issue Summary:

We have two pipelines in a given Kubernetes service - "Deploy" and "Clean Up." "Clean Up" gets triggered on any completion of "Deploy" - success or failure. However, if "Clean Up" is running when "Deploy" gets triggered by a webhook, it may start... and "Clean Up" may clean things up right out from under "Deploy."

While you can specify that multiple instances of a single pipeline shouldn't be allowed to run, it would be nice to have the ability to block one pipeline if another is running. Alternatively, having some sort of "try/finally" style behavior from a given pipeline would work.

This is based on [my StackOverflow question here](https://stackoverflow.com/questions/54658265/can-you-stop-one-spinnaker-pipeline-based-on-a-different-pipeline-running) where it sounds like such a feature doesn't yet exist.

### Cloud Provider(s):

Azure

### Environment:

Spinnaker deploying containers into Azure Kubernetes Service as part of a continuous deployment pipeline.

### Description:

Currently there is no way to have a global error (or success) handler in a given pipeline; I'd love to have that ability.

### Steps to Reproduce:

1. Create a simple pipeline called "Deploy" that may succeed or fail. For testing purposes, this may be a manual judgment.
2. Create a second pipeline called "Clean Up" that runs on all conditions when "Deploy" terminates. Add a manual judgment here, too, so it takes time to complete.
3. Start the "Deploy" pipeline. Cause it to fail. This should kick off the "Clean Up" pipeline.
4. While the "Clean Up" pipeline is still running (e.g., "still cleaning up"), start the "Deploy" pipeline again.
5. Before "Deploy" has reached any sort of terminal state, let "Clean Up" finish.

If "Deploy" is deploying a service and "Clean Up" is handling things like ensuring the network and deployed set of containers is in a good state (e.g., no canary artifacts out there remaining) then it's possible for "Deploy" to push out a new canary... and have "Clean Up" remove all that from underneath the ongoing deployment.

### Additional Details:

There's a lot of detail [in my SO question](https://stackoverflow.com/questions/54658265/can-you-stop-one-spinnaker-pipeline-based-on-a-different-pipeline-running) but I'll do my best to summarize here.

"Deploy" handles all the deployment, testing, canary analysis, etc. for pushing the service out the door. "Clean Up" handles ensuring the network configuration is set to only point at production deployed containers, removes any canary/non-prod versions of the containers, and ensures things are all ready to go for the next deployment.

"Deploy" is a reasonably large pipeline that has around 20 steps. Some of these steps are tests that run in parallel, some are automated canary analysis. If any of these things fail - from baking the Kubernetes manifest to failing to deploy to failing a test - we want to clean up to ensure there aren't any remnants out there.

From a pseudocode perspective, we really want `try/finally` in a pipeline:

```
try {
  // Do the deployment steps.
} finally {
  // Clean things up for the next deployment.
}
```

Putting all the steps from "Clean Up" inline into the "Deploy" pipeline isn't really feasible. There are probably five steps in "Clean Up" and copy/pasting them a bunch isn't awesome. Even if it was just one step that called into the "Clean Up" pipeline, we'd need to do a sort of "catch" around any possible failure and call out to the "Clean Up" pipeline. It seemed that the "trigger a pipeline to run if this other pipeline completes on error" was pretty well fit for this situation.

That said, depending on the timing of when a new build of the service comes in, how long it takes to clean up for the next deployment, etc... the pipeline doing the clean up may run while a new deploy has started, which can cause the new deploy to break.

I didn't see any sort of pipeline step to block if a particular condition was true until said condition was false; nor did I see any sort of workflow mechanism that emulated the `try/finally` or `try/catch/finally` behavior I was looking for. I also wasn't really able to figure out how to fail the "Deploy" pipeline (shoudl something go amiss) yet still call "Clean Up" - it got really, really complicated trying to make a sort of branching construct where, for any failure, all roads would lead to "Clean Up."

By way of comparison, build systems like Azure DevOps (or others that define YAML style build pipelines) [generally have a sort of concept like "conditions"](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml) where I can specify that a particular task should _always_ run or maybe only run on failure.

```yaml
jobs:
- job: Deploy
  steps:
  # a bunch of steps

- job: CleanUp
  steps:
  - script: echo Clean Up
    condition: always() # this step will always run, even if the pipeline fails or is cancelled
```

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by tillig** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by tillig** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by tillig** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by tillig** [comment]
@spinnakerbot remove-label stale

**Comment by NasAmin** [comment]
I am interested in this too but slightly different scenario
We have a number of microservices and each of them have their own deployment pipeline.
We also have staging tests pipeline where all the microservices are tested in a staging cluster.
Whilst the staging tests are running, a microservice pipeline could be triggered by code changes. This means whilst the staging tests are in progress, a new version of a given microservice might be deployed which breaks the integrity of staging tests.

So we'd like to block deployment of micro services whilst they are under test and once the tests are finished then the microservice's pipeline should trigger as normal.

**Comment by robzienert** [comment]
cc @marchello2000 - weren't we talking a bit last week about similar functionality?

**Comment by marchello2000** [comment]
@robzienert : yes, I am actually working on something like this. Will update this issue with I have the PR ready

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by ivan-samartino** [comment]
@marchello2000 , any update on the PR you mentioned?

**Comment by spinnakerbot** [comment]
This issue is tagged as 'to-be-closed' and hasn't been updated  in 45 days, so we are closing it. You can always reopen this issue if needed.

**Comment by mkuziemko** [comment]
The same issue occurred in my case. Are there any updates or plans regarding implementation of this functionality?

**Comment by billcchung** [comment]
I do hope this can be done, we want something that cleans things up even the pipeline is cancelled, currently no further stage can be executed after job is cancelled. We're thinking about a workaround to execute another pipeline to do the clean up tho.

**Comment by airstandley** [comment]
@spinnakerbot remove-label to-be-closed

**Comment by airstandley** [comment]
I have a number of use cases where blocking execution of pipeline A if pipeline B is running would be useful.
Can this be re-opened?

**Comment by amitmahbubani** [comment]
@spinnakerbot remove-label stale
